pantheon-decoupled.php

```diff
@@ -448,6 +467,128 @@ function pantheon_decoupled_redirect_to_fes() {
+  if ( ! current_user_can( 'manage_options' ) ) {
+    wp_die( esc_html__( 'You do not have sufficient permissions to access this page.', 'wp-decoupled-preview' ) );
+  }
+  check_admin_referer( 'env-regen', 'nonce' );
+
+  // Get data for preview site
+  $id = isset( $_GET['id'] ) ? absint( sanitize_text_field( $_GET['id'] ) ) : NULL;
+  $preview_sites = get_option( 'preview_sites' );
+  $preview_site = isset( $preview_sites['preview'][ $id ] ) ? $preview_sites['preview'][ $id ] : NULL;
+  
+  $docs_url = 'https://docs.pantheon.io/guides/decoupled/overview'
+    
+  ?>
+      <style>
+        /* Hide admin bar and padding on top of page. */
+        html.wp-toolbar {
+          padding-top: 0;
+        }
+        #wpadminbar {
+          display: none;
+        }
+      </style>
+      <div class="wrap">
+          <h1><?php esc_html_e( 'Regenerate Environment Variables', 'wp-pantheon-decoupled' ); ?></h1>
+          <?php
+        if ( $id ) {
+          $site_label = $preview_site['label'];
+          $url = wp_nonce_url( add_query_arg( [
+            'page' => 'env_regen_action',
+            'id' => $id,
+          ], admin_url( 'options-general.php' ) ), 'env-regen', 'nonce' );
+          ?>
+          <a id="regen-password" class="button-secondary button-large" href="<?php echo esc_url( $url ); ?>">
+            <?php
+            echo esc_html(
+              // Translators: %s is the preview site label.
+              sprintf( __( 'Regenerate %s WP_APPLICATION_PASSWORD', 'wp-decoupled-preview' ), $site_label )
+            );
+            ?>
+          </a>
+          <p>Checkout the Pantheon
+            <a id="docs-link" target="_blank" href="<?php echo esc_url( $docs_url ); ?>">
+              platform docs
+            </a> for more information about Front-End Site configuration.
+          </p>
+          <?php
+        }
+        ?>
+      </div>
+  <?php
+}
+
+/**
+ * Helper to get password by name.
+ *
+ * @return string
+ */
+function pantheon_decoupled_get_user_application_password( $user_id, $id ) {
+	$passwords = WP_Application_Passwords::get_user_application_passwords( $user_id );
+
+	foreach ( $passwords as $password ) {
+		if ( $password['name'] === strval($id)) {
+			return $password;
+		}
+	}
+	return null;
+}
+
+function pantheon_decoupled_regen_env_vars_action() {
+  if ( ! current_user_can( 'manage_options' ) ) {
+    wp_die( esc_html__( 'You do not have sufficient permissions to access this page.', 'wp-decoupled-preview' ) );
+  }
+  check_admin_referer( 'env-regen', 'nonce' );
+  $id = isset( $_GET['id'] ) ? absint( sanitize_text_field( $_GET['id'] ) ) : NULL;
+  $user_id = get_current_user_id();
+  // If a password exists for this preview site and user, delete it so that a new one can be generated
+  if(WP_Application_Passwords::application_name_exists_for_user($user_id, $id)) {
+    $currentPassword = pantheon_decoupled_get_user_application_password( $user_id, $id );
```

Comment:
My assumptions going into this ticket were that:
* Since a user could have multiple existing application passwords, we couldn't really determine which one was being used.
* There would be no way to get the un-hashed password.

Based on that I was thinking we'd have to treat this more like a token and regenerate a new one every time.

This is a creative solution to the first one. And might imply that we can get the un-hashed password. Does get_user_application_password() make that possible?

Deleting an application password worries me a bit. But knowing the application password for the preview site has benefits. This seems like something to discuss with the team during the meeting tomorrow. If we do decide to use this approach to associate an application password with a preview site, we should make the name of the application password much more unique and explicit - maybe something like 'Password for Preview Site <Label>' And we should follow the same format with what we're now creating as 'Example application password'.